File: ponio/examples/exp_splitting.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include <filesystem>
#include <string>

#include <ponio/observer.hpp>
#include <ponio/problem.hpp>
#include <ponio/runge_kutta.hpp>
#include <ponio/solver.hpp>
#include <ponio/splitting.hpp>
#include <ponio/user_defined_method.hpp>

// solve $\dot{u} = u$ with $u(t=0) = 1$, and $t\in[0,2]$.

int
main()
{
    std::string const dirname = "exp_splitting_data";

    double lambda = 0.3;

    auto f1 = [=]( double, double u )
    {
        return lambda * u;
    };

    auto f2 = [=]( double, double u )
    {
        return ( 1.0 - lambda ) * u;
    };

    auto pb = ponio::make_problem( f1, f2 );

    auto exact_solver_f1 = [=]( auto /* f */, double tn, double yn, double dt ) -> std::tuple<double, double, double>
    {
        return { tn + dt, std::exp( lambda * dt ) * yn, dt };
    };

    auto exact_solver = [=]( auto /* f */, double tn, double yn, double dt ) -> std::tuple<double, double, double>
    {
        return { tn + dt, std::exp( dt ) * yn, dt };
    };

    double const y0 = 1.0;
    double const dt = 0.5;

    // 3 examples
    {
        auto strang = ponio::splitting::make_strang_tuple( std::make_pair( ponio::make_user_defined_method( exact_solver_f1 ), dt ),
            std::make_pair( ponio::runge_kutta::rk_22_ralston(), 0.5 * dt ) );

        auto filename = std::filesystem::path( dirname ) / "exp_strang.dat";
        observer::file_observer fobs( filename );
        ponio::solve( pb, strang, y0, { 0., 5.0 }, dt, fobs );
    }
    {
        auto exact = ponio::make_user_defined_method( exact_solver );

        auto filename = std::filesystem::path( dirname ) / "exp_exact.dat";
        observer::file_observer fobs( filename );
        ponio::solve( pb, exact, y0, { 0., 5.0 }, dt, fobs );
    }
    {
        auto rk2 = ponio::runge_kutta::rk_22_ralston();

        auto filename = std::filesystem::path( dirname ) / "exp_rk2.dat";
        observer::file_observer fobs( filename );
        ponio::solve( pb, rk2, y0, { 0., 5.0 }, dt, fobs );
    }

    return 0;
}

File: ponio/examples/exp_splitting.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import argparse

import matplotlib.pyplot as plt
import numpy as np

name = "exp_splitting"
data_dir = name+"_data"
img_dir = data_dir

parser = argparse.ArgumentParser(
    description=f"Compile, launch and plot results of example `{name}`")
parser.add_argument('--only-save', action='store_true',
                    help="Just save output in img directory")

arguments = parser.parse_args()

if arguments.only_save:
    img_dir = os.path.join("img", name)
    os.makedirs(img_dir, exist_ok=True)


make = subprocess.Popen(["make", name])
make.wait()

args = [os.path.join(".", name)]
process = subprocess.Popen(args)
process.wait()

test_cases = [
    ("strang", "Strang splitting with exact solver and RK(2,2) Ralston"),
    ("exact", "exact solver"),
    ("rk2", "RK(2,2) Ralston")
]

for tag, label in test_cases:
    data = np.loadtxt(os.path.join(data_dir, f"exp_{tag}.dat"))

    plt.plot(data[:, 0], data[:, 1], "-+", label=label)

plt.plot(data[:, 0], np.exp(data[:, 0]), "--", label="exact solution")
plt.legend()

plt.savefig(os.path.join(img_dir, "01.png"))

if not arguments.only_save:
    plt.title(r"$\dot{u} = \lambda u + (1-\lambda) u$, $u(t=0)=1$")
    plt.show()
